### ISSUE TITLE
DX: Pipeline Synchronization

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#4205]
UPDATE: New proposals with alternative solutions for the 3 problems described here have been created to facilitate more focused discussions:
- #5065
- #5083 
- #4928

---

I've noticed, either while dogfooding or looking at community demos, that the DX is lacking in synchronization mechanism and people are building workarounds around that.

## Problem 1: Single pipeline synchronization

Sometimes, you want to synchronously wait for something to complete

**Current workaround**: Call a synchronous function and throw away the result.

```go
, err = ctr.ExitCode(ctx)
, err = dir.Entries(ctx)
```

Proposed solution: Explicit `Wait()` function

```go
ctr.Wait(ctx)
dir.Wait(ctx)
```

## Problem 2: Multiple pipeline synchronization

Often times, you want to execute multiple pipelines in parallel (e.g. build & test). This is surprisingly tedious to do.

- **Workaround 1**: Use a "join pipeline" + the workaround described for single pipelines

```go
// Both return a `dagger.Container`, full async
testResult := test()
lintResult := lint()

// Do *something* that forces a wait on both
_, err := c.Container().From("alpine").
  WithMountedDirectory("/a", testResult).
  WithMountedDirectory("/b", lintResult).
  WithExec([]string{"ls"}).
  ExitCode(ctx)
return err
```

- **Workaround 2**: Use language parallelization + above workaround

```go
// execute synchronous functions in go routines
eg, gctx := errgroup.WithContext(ctx)
eg.Go(func() error {
  _, err := test().ExitCode(gctx)
  return err
})

eg.Go(func() error {
  _, err := lint().ExitCode(gctx)
  return err
})

return eg.Wait()
```

**Proposed solution**:

Introduce a `dagger.Wait()` synchronization helper:

```go
return dagger.Wait(
  test(),
  lint(),
)
```

## Problem 3: Orphaned pipelines

If a synchronous function is not explicitly called, the entire pipeline is orphaned:

```go
lint()

testOutput, err := test().Stdout(ctx)
```

In the above example, `lint` is **never** executed.

**Current workaround**: Don't forget 

**Proposed solution**: SDKs will automatically call `Wait()` on all pipelines upon session completion, forcing execution of "forgotten" pipelines

**Comment by shykes** [comment]
I like the primitive, but not sure about the name. `Wait` implies that something is already happening, and you want to wait for it to complete. But this primitive does more than that: it ensures that something actually happens, when otherwise it wouldn't. That's pretty different from eg. the Unix `wait()`.

I don't have a better name to offer though, sorry. Will try to think of alternatives.

**Comment by helderco** [comment]
Same. Like the idea but don't have a better name:

```python
await ctr.wait()  # 🤪
```

**Comment by helderco** [comment]
How about `dagger.Do()`?

```python
await dagger.do(ctr, test())
```

**Comment by slumbering** [comment]
> Same. Like the idea but don't have a better name:
> 
> ```python
> await ctr.wait()  # 🤪
> ```

Same ambiguity for Node.js.

**Comment by vito** [comment]
I like the idea of trying to eliminate this pattern (wanted to open an issue for it, but found this one):

```go
code, err := ctr.ExitCode(ctx)
if err != nil {
  return err
}

if code != 0 {
  return fmt.Errorf("container failed")
}
```

it'll appear in just about everyone's CI pipelines for running tests/smoke tests/other kinds of checks that don't have direct downstream dependents and only serve to gatekeep the pipeline.

Right now the last 3 lines are redundant, but that's a bug (#3192). We shouldn't encourage `_, err := ctr.ExitCode(ctx)` because people will likely have to care about that exit code once that bug is fixed. Fixing the bug would mean all these cases silently ignore failures.

> **Proposed solution**:
> 
> Introduce a `dagger.Wait()` synchronization helper:

I prefer leaving this to the language runtime (option 2), but I'm not 100% sure. My motivation for that is to keep SDKs light and stay true to the promise of "pipelines as regular code" so people can use whatever concurrency primitives they normally do.

I think my ideal would be something like `(Container).Run(context.Context) error`. Then you'd do something like this:

```go
// execute synchronous functions in go routines
eg, gctx := errgroup.WithContext(ctx)
eg.Go(func() error {
  return test().Run(gtcx)
})

eg.Go(func() error {
  return lint().Run(gtcx)
})

return eg.Wait()
```

Even a theoretical Bash SDK should be able to handle concurrency:

```sh
dagger do test &
dagger do lint &
wait
```

**Comment by marcosnils** [comment]
> I prefer leaving this to the language runtime (option 2), but I'm not 100% sure. My motivation for that is to keep SDKs light and stay true to the promise of "pipelines as regular code" so people can use whatever concurrency primitives they normally do.

+1. We (or the community) could build utility helpers around the SDK's to support this when needed. Following `golang` history, packages like `errgroup` weren't introduced in the language until they went through some time of the community using them and giving feedback. If we do have the necessity to use this today, we could create a new Go lib which helps until we validate it's what users really need.

**Comment by sagikazarmark** [comment]
> Problem 3: Orphaned pipelines

I have mixed feelings about this. On one hand, if I fail to notice that something is not executed, that could be problematic.

On the other hand, I may want to leave certain containers out on purpose. Yes, I could just avoid creating them, but I kinda like knowing that they are lazily evaluated and only those are executed that become part of the final graph.

**Comment by robertu94** [comment]
I think it should be preferred to leave synchronization to the language SDK, but ensure that is is marked in such a way that tooling helps catch these errors.  This means one less thing for users to learn and internalize when picking up docker.

So what do I mean by tooling helps catch these errors?  In languages such as C++ and Rust, it is possible to mark a struct as `[[nodiscard]]` or `#[must_use]`, and the ways to create Future objects have this attribute.  This means if a function returns an object of this type, the calling code must use it.  That doesn't mean that the result must be acquired, but it must either be consumed with the the `.get()` method or it's equivalent, be explicitly canceled, or returned from the function with same restrictions.  Python's asyncio.run has an option `debug=True` which enables similar warnings.  I don't use go very often, but I believe `go vet` has a `unusedresult` warning I would expect should/could be made to catch this.  I also don't use Javascript very often, but ESlint also has a `no-floating-promises` which also seems to do similar things.

**Comment by helderco** [comment]
I think we should bump the priority on figuring out a solution for this. There's lots of workarounds and confusion because we don't have this.

**Comment by helderco** [comment]
> I prefer leaving this to the language runtime (option 2), but I'm not 100% sure. My motivation for that is to keep SDKs light and stay true to the promise of "pipelines as regular code" so people can use whatever concurrency primitives they normally do.

I think it's good if you still have a way to use your language's primitives if you want, but still allow using a convenience where the SDK does it for you. I've seen a lot of people not reaching correctly for concurrency. It can make a bit impact on performance, not sure if people are using it enough.

- #4766 

Even in Go, where it's considered easy enough to use, there's caveats to be aware of and so without a good understanding, it can become a burden to depend on that.

TLDR: We could hide away the concurrency in SDK conveniences like this, but still allow an advanced user to use the language's primitives for more granular control.

**Comment by helderco** [comment]
> > Problem 3: Orphaned pipelines
> 
> I have mixed feelings about this. On one hand, if I fail to notice that something is not executed, that could be problematic.
> 
> On the other hand, I may want to leave certain containers out on purpose. Yes, I could just avoid creating them, but I kinda like knowing that they are lazily evaluated and only those are executed that become part of the final graph.

We can just log a warning instead, silence-able by a setting or environment variable if too noisy.

**Comment by helderco** [comment]
> This means if a function returns an object of this type, the calling code must use it. That doesn't mean that the result must be acquired, but it must either be consumed with the the `.get()` method or it's equivalent, be explicitly canceled, or returned from the function with same restrictions. 

Are you thinking more of "orphaned pipelines", or all of this issue's proposal? In Python, you get a warning if you ignore a coroutine (e.g., not using `await`). Fields like `exit_code` or `stdout` return a coroutine so the SDK is already flagging those as "must use".

That's different from having a dynamic synchronization method to force executing an otherwise lazy pipeline, without resorting to `exit_code` as a common workaround, or executing multiple lazy pipelines concurrently for you.

**Comment by robertu94** [comment]
> > This means if a function returns an object of this type, the calling code must use it. That doesn't mean that the result must be acquired, but it must either be consumed with the the `.get()` method or it's equivalent, be explicitly canceled, or returned from the function with same restrictions.
> 
> Are you thinking more of "orphaned pipelines", or all of this issue's proposal? In Python, you get a warning if you ignore a coroutine (e.g., not using `await`). Fields like `exit_code` or `stdout` return a coroutine so the SDK is already flagging those as "must use".

Yes @helderco , I was principally addressing issue 3 with that comment.  Tooling should help you identify orphaned pipelines.

More broadly, I am advocating for Dagger to use language features and types (e.g. futures, go routines) to handle synchronization.  If Dagger does this, tooling in each language should help detect orphaned pipelines.

**Comment by helderco** [comment]
Just an update: I've been thinking about this as I think it's very important for DX. I need to create a few related issues to keep discussions focused, then I'll update this with how things stand and push for a decision.

**Comment by helderco** [comment]
Looks to me that at least some of us like the general idea, there’s just a few concerns preventing this from moving forward.

I’ve compiled the following list to see if we can untie these knots and move forward on a decision. I believe this is an important DX improvement that will hopeful simplify and clear some confusion linked to using other kind of workarounds.

Anything missing here?

## Concern 1: Name

- OP suggested `.Wait()` but people are [unsure](https://github.com/dagger/dagger/issues/4205#issuecomment-1353817793) about the name
- I suggested [`.Do()`](https://github.com/dagger/dagger/issues/4205#issuecomment-1353893454) instead
- Any more?

Let’s not let bikesheding be the hill this dies on! 😅

## Concern 2: Orphan detection

Some users may want to leave orphans on purpose. I suggested simply raising a warning that can be suppressed.

I created a new issue for that, so we can keep this issue focused on synchronization:
- https://github.com/dagger/dagger/issues/4928

## Concern 3: Replacing `exitCode` 

Not really a concern, but wondering about possibilities here.

We need to replace using `.ExitCode(ctx)` as a pattern, especially since it doesn’t work and people are using it as a “hammer and everything looks like a nail”. Explanation and proposal in:
- https://github.com/dagger/dagger/issues/4921

### `Run()` convenience

From that discussion, there’s talk of a [`.Run()` convenience](https://github.com/dagger/dagger/issues/4921#issuecomment-1503601125) for the “happy path” of not caring about the result, just error on any failure:

```go
err := ctr.Run(ctx)
if err != nil {
    panic(err)
}
```

Can `ctx.Wait(ctx)` replace that?

At first it looks like it, but there’s other caveats in that idea that I want to explore in light of a general `.Wait()`.

### Dockerfile build

Additionally from https://github.com/dagger/dagger/issues/4668#issuecomment-1490289888, we also need a way to trigger a `Dockerfile` build without having to resort to hacks like `Build().WithExec(nil).ExitCode(ctx)` or `Build().Rootfs().Entries(ctx)`.

In this case, some users want to just check if it builds, and don’t need to push to registry or do anything else that would trigger the build.

I assume that `Build().Wait(ctx)` solves this, just requesting for comment if it doesn’t. And any other use case that uses `ExitCode` as a hack, not covered by `.Wait()` for that matter.

## Concern 4: Using the language’s concurrency

Proposed solution is via an SDK helper:

```go
return dagger.Wait(
  test(),
  lint(),
)
```

Some users would rather depend on the language’s features for this, but I want to suggest both: this helper **wouldn’t be a replacement** for using the language’s concurrency, it would be a **convenience** or *syntax sugar*.

I’ve seen users shy away from doing concurrency because they’re just not familiar with it and don’t want to invest time in learning it properly.

I also think a lot of users **aren’t well aware** that using concurrency can make a big difference:
- https://github.com/dagger/dagger/issues/4766

When using an `await` inside an `async` function, some users think that it’s already asynchronous == concurrent, but that’s not true. You need a task group (in Python) or an error group (in Go) to make it concurrent.

Perhaps some users also avoid it because it adds verbosity to their code and they don’t want that boilerplate. Ties back to not being aware that it’s probably worth it.

I think reducing the “concurrency burden” on the user as much as possible, via an SDK helper for example, is definitely helpful (no pun intended 😁).

One example where we’re already doing this is when you call a `Container.stdout` with multiple mounts, those lazy `Directory` objects are concurrently resolving into corresponding IDs, before sending the `stdout` query[^1].

[^1]: At least in Python, I haven’t checked the other SDKs’ implementations.

In that case we decided on avoiding the ability to use IDs directly for usability, but in this case (synchronization) I don’t think anyone wants to prevent being able to use goroutines/coroutines/promises directly, especially since you may need to do something more advanced.

It’s also useful just to reduce the boilerplate and make pipeline code simpler in (probably) most cases, including making it easier to port code between SDKs.

Again, this would be just a **convenience**.

**Comment by gerhard** [comment]
This keeps coming up in various conversations. Addressing it is important for moving forward.

Having talked to @helderco about this, the primary blocker for moving this forward is the naming. We spoke about various names synchronously, went back-and-forth, and both settled on `Sync()`. The [`sync(1)`](https://manpages.ubuntu.com/manpages/jammy/en/man1/sync.1.html) semantics seem a natural fit for this context. There is a strong hint that this is the right name from the issue title, as well as how we talk about this problem in various comments.

In the interest of moving this forward, if none disagrees with `Sync()` by May 1st, we would like to get this resolved via a new PR next week.

cc @shykes @aluzzardi @vito @marcosnils @sagikazarmark @robertu94 

---

Worth pointing out, this doesn't close the door on any of the concerns or ideas others mentioned. A few weeks ago, @helderco created a bunch of proposals for your concerns - see the link references a few comments above.

Most importantly, anyone will be able to leverage their preferred language concurrency primitives, even when using `Sync()`.

**Comment by d3rp3tt3** [comment]
@mircubed

**Comment by shykes** [comment]
> In the interest of moving this forward, if none disagrees with `Sync()` by May 1st, we would like to get this resolved via a new PR next week.

I don't know exactly what proposal you're asking for a decision on?

Currently I see this in the top comment:

Part 1:

> Explicit Wait() function
> ```
> ctr.Wait(ctx)
> dir.Wait(ctx)
> ```

Part 2:

> Introduce a dagger.Wait() synchronization helper:
> ```
> return dagger.Wait(
 > test(),
>  lint(),
> )
> ```

Part 3:

> SDKs will automatically call Wait() on all pipelines upon session completion, forcing execution of "forgotten" pipelines

Part 4:

>  if none disagrees with `Sync()`

I have several questions:

- What parts of the proposal are renamed to `Sync`? Only part 1, only part 2, or both?
- What are the types in part 1 and 2? Is this for Container only, or several types?
- Which parts of this proposal are SDK-specific? Are there any new GraphQL API changes?

**Comment by helderco** [comment]
I'm working on updating this with a full proposal, just to quickly answer your questions.

> Part 3:
>
> > SDKs will automatically call Wait() on all pipelines upon session completion, forcing execution of "forgotten" pipelines

In my last comment I suggested we abandon this in favor of another proposal, just for this point:
- https://github.com/dagger/dagger/issues/4928

> What parts of the proposal are renamed to Sync? Only part 1, only part 2, or both?

Both. The second is just calling the first on multiple lazy objects concurrently.

> What are the types in part 1 and 2? Is this for Container only, or several types?

Several types, but I've yet to make a list. Whenever it can be useful to force an evaluation.

> Which parts of this proposal are SDK-specific? Are there any new GraphQL API changes?

I'm thinking part 1 in the API (using an interface with a field for this purpose), part 2 in the SDKs (single query using the query builder or using the language's concurrency in a loop).

I intend to put the spec in the proposal.

**Comment by busla** [comment]
borrowing async syntax conventions from C#, JS, and eventually adopted by Python and Rust, might be helpful to streamline things between the Dagger SDKs. I'm not a Go expert (a total noob to be honest) so maybe this is out of wack, but maybe something like this? 

```go
func main() {
	ctx := context.Background()
	client, err := dagger.Connect(ctx)
	if err != nil {
		panic(err)
	}
	defer client.Close()

	futureContainer := client.Pipeline("test").
		Container().
		From("alpine").
		WithExec([]string{"apk", "add", "curl"}).
		WithExec([]string{"curl", "https://dagger.io"}).Async() // returns a promise
	
	container := futureContainer.Await()
}
```

**Comment by helderco** [comment]
@busla, that doesn't seem worth it (also not a Go expert 😅).

In Python and Node.js the `async`/`await` get out of the way, being actually syntax sugar for more verbose code (e.g., in Node.js it simplifies the `Promise().then()` syntax).

In the non-Go SDKs, we only return "promises" on leaf values, i.e., when we want to make the GraphQL request. So the equivalent would be more like:

```go
out, err := client.Pipeline("test").
	Container().
	From("alpine").
	WithExec([]string{"apk", "add", "curl"}).
	WithExec([]string{"curl", "https://dagger.io"}).
	Stdout().  // returns a generic Future[string]
	Await(ctx)
```

So this adds another step. Instead of just `.Stdout(ctx)` it's now `.Stdout().Await(ctx)`, and it's still not enough to streamline like you said. Doing that in Go would add complexity without much benefit.

We've also designed our SDKs with **idiomatic** code in mind.

Anyway, this discussion is more suited to:
- #4668 

It was actually the kind of thing we discussed in:
- #3555
